### PR TITLE
Merge the net10 runtime, aspnetcore, windowsdesktop, and arcade codeflows

### PR DESCRIPTION
**Pull request by v-wuzhai**
Merge #43061 and #42842

### CODE REVIEW

**Comment by v-wuzhai** [comment]
@marcpopMSFT I attempted to merge the net10 dotnet/runtime and dotnet/aspnetcore codeflows, manually updating some dependencies for 10.0 and adding 10.0 templates. The build succeeded, but many tests failed. The failure information is as follows: 
`System.TypeLoadException : Could not load type 'System.Runtime.InteropServices.MemoryMarshal' from assembly 'System.Runtime, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.`

Could you look at this?

**Comment by marcpopMSFT** [comment]
VMR leg failed with the following (@dotnet/product-construction )

- Test failures in Ubuntu
  - CompareAssemblyVersions and CompareFileContents
- Windows leg failed hte build
  - D:\a\_work\1\vmr\repo-projects\Directory.Build.targets(476,5): error : Repository asset manifest files don't exist. [D:\a\_work\1\vmr\repo-projects\fsharp.proj]

**Comment by marcpopMSFT** [comment]
Source build leg failed with a prebuilt for net10 (@dotnet/source-build-internal )
__w/1/s/.packages/microsoft.dotnet.arcade.sdk/10.0.0-beta.24430.1/tools/SourceBuild/AfterSourceBuild.proj(81,5): error : 1 new pre-builts discovered! Detailed usage report can be found at /__w/1/s/artifacts/sb/prebuilt-report/baseline-comparison.xml.
 Microsoft.NETCore.App.Crossgen2.linux-x64.10.0.0-alpha.1.24431.5

**Comment by marcpopMSFT** [comment]
`System.TypeLoadException : Could not load type 'System.Runtime.InteropServices.MemoryMarshal' from assembly 'System.Runtime, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.`
> 
> Could you look at this?

@dotnet/dnr-codeflow any ideas on this one? Seems to be the common failure across most of the tests I scanned through.

**Comment by marcpopMSFT** [comment]
@v-wuzhai were you going to start on the changes for GenerateBundledVersions.targets for the known framework references and known runtime packs (amongst others)? I think that change can be prepped in parallel to this:
I believe you did this last time here: https://github.com/dotnet/installer/pull/17432

**Comment by NikolaMilosavljevic** [comment]
> @v-wuzhai were you going to start on the changes for GenerateBundledVersions.targets for the known framework references and known runtime packs (amongst others)? I think that change can be prepped in parallel to this: I believe you did this last time here: [dotnet/installer#17432](https://github.com/dotnet/installer/pull/17432)

The change was made with https://github.com/dotnet/sdk/pull/42969

It is now causing 2nd-stage failures in VMR: https://dev.azure.com/dnceng/internal/_build/results?buildId=2531460&view=logs&j=9f4c4083-9864-5575-dedd-fb296084d8e2&t=15f0c853-f021-5549-d93c-f01428803354

```
/vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Ref with version (= 9.0.0-preview.7.24405.7)
/vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102:   - Found 8 version(s) in /tmp/tmp.xLJpd9ApwU [ Nearest version: 9.0.0-rc.1.24414.5 ]
/vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102:   - Found 0 version(s) in /vmr/.dotnet/library-packs
```

**Comment by MichaelSimons** [comment]
> > @v-wuzhai were you going to start on the changes for GenerateBundledVersions.targets for the known framework references and known runtime packs (amongst others)? I think that change can be prepped in parallel to this: I believe you did this last time here: [dotnet/installer#17432](https://github.com/dotnet/installer/pull/17432)
> 
> The change was made with #42969
> 
> It is now causing 2nd-stage failures in VMR: https://dev.azure.com/dnceng/internal/_build/results?buildId=2531460&view=logs&j=9f4c4083-9864-5575-dedd-fb296084d8e2&t=15f0c853-f021-5549-d93c-f01428803354
> 
> ```
> /vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Ref with version (= 9.0.0-preview.7.24405.7)
> /vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102:   - Found 8 version(s) in /tmp/tmp.xLJpd9ApwU [ Nearest version: 9.0.0-rc.1.24414.5 ]
> /vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102:   - Found 0 version(s) in /vmr/.dotnet/library-packs
> ```

Should the hardcoded value be changed to a property that can be overridden by source-build so that it can pick up the version from previous artifacts?

**Comment by NikolaMilosavljevic** [comment]
> > > @v-wuzhai were you going to start on the changes for GenerateBundledVersions.targets for the known framework references and known runtime packs (amongst others)? I think that change can be prepped in parallel to this: I believe you did this last time here: [dotnet/installer#17432](https://github.com/dotnet/installer/pull/17432)
> > 
> > 
> > The change was made with #42969
> > It is now causing 2nd-stage failures in VMR: https://dev.azure.com/dnceng/internal/_build/results?buildId=2531460&view=logs&j=9f4c4083-9864-5575-dedd-fb296084d8e2&t=15f0c853-f021-5549-d93c-f01428803354
> > ```
> > /vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Ref with version (= 9.0.0-preview.7.24405.7)
> > /vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102:   - Found 8 version(s) in /tmp/tmp.xLJpd9ApwU [ Nearest version: 9.0.0-rc.1.24414.5 ]
> > /vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102:   - Found 0 version(s) in /vmr/.dotnet/library-packs
> > ```
> 
> Should the hardcoded value be changed to a property that can be overridden by source-build so that it can pick up the version from previous artifacts?

Yes, that is my idea for a fix, but I need to confirm that this is the case, in a local repro.

**Comment by NikolaMilosavljevic** [comment]
> > > > @v-wuzhai were you going to start on the changes for GenerateBundledVersions.targets for the known framework references and known runtime packs (amongst others)? I think that change can be prepped in parallel to this: I believe you did this last time here: [dotnet/installer#17432](https://github.com/dotnet/installer/pull/17432)
> > > 
> > > 
> > > The change was made with #42969
> > > It is now causing 2nd-stage failures in VMR: https://dev.azure.com/dnceng/internal/_build/results?buildId=2531460&view=logs&j=9f4c4083-9864-5575-dedd-fb296084d8e2&t=15f0c853-f021-5549-d93c-f01428803354
> > > ```
> > > /vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Ref with version (= 9.0.0-preview.7.24405.7)
> > > /vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102:   - Found 8 version(s) in /tmp/tmp.xLJpd9ApwU [ Nearest version: 9.0.0-rc.1.24414.5 ]
> > > /vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj : error NU1102:   - Found 0 version(s) in /vmr/.dotnet/library-packs
> > > ```
> > 
> > 
> > Should the hardcoded value be changed to a property that can be overridden by source-build so that it can pick up the version from previous artifacts?
> 
> Yes, that is my idea for a fix, but I need to confirm that this is the case, in a local repro (building atm).

Wouldn't that solution break the build when runtime starts producing 10-versioned App.Ref package? All projects targeting 9.0 would need to be updated to target 10.0, or we'd need a 9.0 SBRP package - but only after 9.0 RTM.

**Comment by MichaelSimons** [comment]
> Wouldn't that solution break the build when runtime starts producing 10-versioned App.Ref package? All projects targeting 9.0 would need to be updated to target 10.0, or we'd need a 9.0 SBRP package - but only after 9.0 RTM.

Correct, Once 9.0 GAs, we would add the 9.0 targeting pack to SBRP and this issue goes away.

**Comment by marcpopMSFT** [comment]
Interesting, I have a special rule for mentions but just adding me as a reviewer just goes in my clutter folder. So I had no idea that work had already been done. I think getting this working is the next step. Still need someone from runtime to look at the MemoryMarshal exception in the tests.

**Comment by v-wuzhai** [comment]
> `System.TypeLoadException : Could not load type 'System.Runtime.InteropServices.MemoryMarshal' from assembly 'System.Runtime, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.`
> 
> > Could you look at this?
> 
> @dotnet/dnr-codeflow any ideas on this one? Seems to be the common failure across most of the tests I scanned through.

@dotnet/runtime-infrastructure Could you take a look at the failures here?

**Comment by agocke** [comment]
My attempts to replicate the VMR build here are not working. Could someone link to instructions on how to diagnose this?

**Comment by lewing** [comment]
cc @akoeplinger

**Comment by MichaelSimons** [comment]
> My attempts to replicate the VMR build here are not working. Could someone link to instructions on how to diagnose this?

One "easy" way is to utilize this devcontainer - https://github.com/dotnet/sdk/blob/main/.devcontainer/vmr/README.md

**Comment by steveisok** [comment]
@joeloff part of the VMR win-x64 failure appears to be wix related

**Comment by joeloff** [comment]
> @joeloff part of the VMR win-x64 failure appears to be wix related
Which leg?

**Comment by marcpopMSFT** [comment]
@steveisok where do you see wix?

**Comment by steveisok** [comment]
> > @joeloff part of the VMR win-x64 failure appears to be wix related
> > Which leg?

VMR Vertical Build Windows_x64

**Comment by marcpopMSFT** [comment]
For the SDK build, we have 4 container test failures and a blazor failure @lewing @baronfel 
`\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(495,5): error NETSDK1112: The runtime pack for Microsoft.NETCore.App.Runtime.Mono.browser-wasm was not downloaded.`
`System.InvalidOperationException : The registry was not loaded after 5 retries. Spawn registry attempt #1 failed, Xunit.Sdk.XunitException: Expected command to pass but it did not.`

**Comment by marcpopMSFT** [comment]
> VMR Vertical Build Windows_x64 - windowsdesktopbuild

Ahh, it has more than one error. When you click on the build, it takes you to the aspnet failure which you linked above but if you scroll, you get a wix failure. I didn't realize they would have multiple errors that far apart in the log.

**Comment by steveisok** [comment]
> Ahh, it has more than one error. When you click on the build, it takes you to the aspnet failure which you linked above but if you scroll, you get a wix failure. I didn't realize they would have multiple errors that far apart in the log.

In a lot of cases the build will not stop when one repo or another has an error.

**Comment by joeloff** [comment]
Looks like the failure happens when trying to link the desktop bundle, and that's why there already are retries

https://github.com/dotnet/arcade/blob/344004446883e2d577d21a9844f997215964279a/src/Microsoft.DotNet.Build.Tasks.Installers/build/wix/wix.targets#L348C5-L359C28

**Comment by lewing** [comment]
> For the SDK build, we have 4 container test failures and a blazor failure @lewing @baronfel `\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(495,5): error NETSDK1112: The runtime pack for Microsoft.NETCore.App.Runtime.Mono.browser-wasm was not downloaded.` `System.InvalidOperationException : The registry was not loaded after 5 retries. Spawn registry attempt #1 failed, Xunit.Sdk.XunitException: Expected command to pass but it did not.`

given that is this includes the version switch from 9 to 10 I imagine the workload update logic is confused

**Comment by marcpopMSFT** [comment]
Checking on status:

- @lewing what's the fix for the browser-wasm failure if it is a migration from 9 to 10 causing it
- @baronfel I assume the container tests are the known failure and can be ignored if it comes to that. Should we add the test disable commit to this PR perhaps?
- @MichaelSimons I saw that https://github.com/dotnet/source-build/issues/4607 was marked as compelted. Any changes needed in this PR?
- @agocke any luck with the remaining VMR failures? Were you looking into those or someone else?

**Comment by MichaelSimons** [comment]
> @MichaelSimons I saw that https://github.com/dotnet/source-build/issues/4607 was marked as compelted. Any changes needed in this PR?

That issue is orthogonal to the failure here.

**Comment by agocke** [comment]
The Linux VMR failures look like version issues that aren't specific to runtime.

The Windows VMR issue, after a little digging, looks a bit strange... I wonder if this is an overload resolution change. @333fred Can you take a look? The specific call it seems to be complaining about is `Expression.Lambda(typeof(Func<object?>), methodCallExpression!)` and the signature it's picking is `System.Linq.Expressions.Expression.Lambda(Expression, params ParameterExpression[])`, which doesn't look right.

**Comment by 333fred** [comment]
> The Windows VMR issue, after a little digging, looks a bit strange... I wonder if this is an overload resolution change. @333fred Can you take a look? The specific call it seems to be complaining about is `Expression.Lambda(typeof(Func<object?>), methodCallExpression!)` and the signature it's picking is `System.Linq.Expressions.Expression.Lambda(Expression, params ParameterExpression[])`, which doesn't look right.

This doesn't seem likely. The error being seen here is from the IL analyzers, not from Roslyn; if overload resolution was actually resolving to `System.Linq.Expressions.Expression.Lambda(Expression, params ParameterExpression[])`, then I'd expect to see a real compiler error. As it is, that signature isn't even applicable to a call site of `System.Type, Expression`, so overload resolution wouldn't even see it. This seems more likely to be some kind of analyzer bug.

**Comment by 333fred** [comment]
To troubleshoot the windows VMR issue further (if it's consistent), I'd need a binlog of the build and the sources that went into it, or more preferrably, a complog.

**Comment by lewing** [comment]
@marcpopMSFT https://github.com/dotnet/runtime/pull/107899 is at least part of the problem but it won't really work on an rc1 sdk so I'll have to look closer

**Comment by marcpopMSFT** [comment]
> To troubleshoot the windows VMR issue further (if it's consistent), I'd need a binlog of the build and the sources that went into it, or more preferrably, a complog.

@dotnet/product-construction anyone know how to get this from the VMR build? The aspnet build just has a normal build log and is missing the binlog in the artifacts tab? Not sure how to get a complog either.

**Comment by akoeplinger** [comment]
You can get the binlog by going to Windows_x64_BuildLogs_Attempt1 > artifacts > log > Release > roslyn > source-inner-build.binlog

**Comment by 333fred** [comment]
@akoeplinger what about getting the sources that went into the build?

**Comment by agocke** [comment]
@jtschuster helped me track this down. It looks like the code in aspnetcore here is missing a fix for these warnings that's in main. If someone brings in a newer version of aspnetcore, that should probably fix it.

**Comment by lewing** [comment]
> @jtschuster helped me track this down. It looks like the code in aspnetcore here is missing a fix for these warnings that's in main. If someone brings in a newer version of aspnetcore, that should probably fix it.

`darc update-dependencies -c ".net 10" --source-repo dotnet/aspnetcore`

**Comment by marcpopMSFT** [comment]
Latest build is down do a single SDK test failure that Larry is looking into.

We still have the VMR/unified build failures and source build failures. 

For unified build, efcore seems to have a lot of package failures.
 
Still the memory marshal failure source build.

**Comment by agocke** [comment]
In the VMR leg it looks like all the versions are broken. I can't find any other failures. Marc can you pick out the error you're referencing?

**Comment by marcpopMSFT** [comment]
Summarizing the various errors I see.

**SDK build test**
Publish_HostedApp_WithRidSpecifiedAsArgument_NoSelfContained_Works -- error `NETSDK1112: The runtime pack for Microsoft.NETCore.App.Runtime.Mono.browser-wasm was not downloaded.` 

**Source Build**
`    /vmr/src/sdk/artifacts/sb/package-cache/microsoft.net.compilers.toolset/4.12.0-3.24463.1/tasks/netcore/Microsoft.CSharp.Core.targets(89,5): error : Unhandled exception. System.TypeLoadException: Could not load type 'System.Runtime.InteropServices.MemoryMarshal' from assembly 'System.Runtime, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. [/vmr/src/sdk/src/Cli/Microsoft.DotNet.InternalAbstractions/Microsoft.DotNet.InternalAbstractions.csproj]`

**unified build**
`    /vmr/src/efcore/src/EFCore.Abstractions/EFCore.Abstractions.csproj : error NU1103: Unable to find a stable package Microsoft.NET.ILLink.Tasks with version (>= 8.0.7) [/vmr/src/efcore/EFCore.sln]`

**Comment by mmitche** [comment]
The efcore issue should be handled by https://github.com/dotnet/efcore/pull/34719, and adding the windowsdesktop flow in here should the issue around WiX.

**Comment by lewing** [comment]
runtime is still blocked on https://github.com/dotnet/runtime/pull/106599 and the perf lane fixes

**Comment by marcpopMSFT** [comment]
@lewing any word on that runtime PR?
@mmitche the VMR results look worse now with clang and emcmake errors
@dotnet/source-build-internal source build is complaining about a prebuilt that looks like it just comes from runtime: System.IO.Hashing.10.0.0-alpha.1.24471.6

**Comment by lewing** [comment]
https://github.com/dotnet/runtime/pull/106599 has merged, will update this pr once there is an official build

**Comment by lewing** [comment]
@ericstj this will also be a good place to look for task/nuget related problems once it is updated

**Comment by lewing** [comment]
Runtime update is here now.

**Comment by lewing** [comment]
@jkoritzinsky some brotli issues in VMR

**Comment by mmitche** [comment]
The arcade has some bug that hits in aspnetcore related to dotnet tool restore on the Unix legs. I am investigating this.

**Comment by lewing** [comment]
I bumped the sourcebuild dotnet and arcade, not sure that was correct but that is where the MemoryMarshal error was coming from

**Comment by mmitche** [comment]
Yeah, I'm not sure why. @dkurepa @premun @MichaelSimons  any ideas?

**Comment by mmitche** [comment]
> @mmitche how does EF Core synchronize? possible to remove the patch and allow the update to flow normally?

via aspnetcore, which is blocked up behind the aspnetcore net10 update /cc @wtgodbe. We may be able to flow efcore separate from, the net10 update. let me look into it.

**Comment by mmitche** [comment]
@kasperk81 Looking into it.

**Comment by nagilson** [comment]
@mmitche What are the next steps to get this PR through?

**Comment by mmitche** [comment]
> @mmitche What are the next steps to get this PR through? Thanks for helping drive this forward.

Investigations of the UB test failures and the last bit of source build failures.

**Comment by MichaelSimons** [comment]
@mmitche - I propose we disable the two failing SB tests and log follow-ups.  The dotnet-watch test is likely dependent on the full net10 transition.  The symbols tests is more questionable but not worth blocking this effort.

**Comment by MichaelSimons** [comment]
> @mmitche - I propose we disable the two failing SB tests and log follow-ups. The dotnet-watch test is likely dependent on the full net10 transition. The symbols tests is more questionable but not worth blocking this effort.

I disabled these two tests.

**Comment by mmitche** [comment]
FINGERS AND TOES CROSSED

File: src/SourceBuild/content/eng/tools/prebuilt-baseline.xml
Comment (MichaelSimons) [inline src/SourceBuild/content/eng/tools/prebuilt-baseline.xml:1]: This file should be removed.  Can be done in a follow-up PR.

**Review by MichaelSimons — APPROVED** [review]

File: src/SourceBuild/content/eng/tools/prebuilt-baseline.xml
Comment (marcpopMSFT) [reply src/SourceBuild/content/eng/tools/prebuilt-baseline.xml:1]: add a comment in https://github.com/dotnet/sdk/issues/42920. 

Anything else from this PR that will need to be unwound?

**Comment by am11** [comment]
Thanks everyone!

For some reason, daily link is still pointing to the August build of runtime: dotnet/runtime@0fbd81404d1f211572387498474063bc6f407f0f

```sh
% curl -sSL https://aka.ms/dotnet/10.0.1xx/daily/productCommit-osx-arm64.json

{
  "runtime": { "commit": "0fbd81404d1f211572387498474063bc6f407f0f", "version": "9.0.0-rc.1.24414.5" },
  "aspnetcore": { "commit": "42af9fe6ddd7c3f9cde04ac003bf97509881873b", "version": "9.0.0-rc.1.24414.4" },
  "windowsdesktop": { "commit": "c3466bfa163ab8f93e92a364dbc2b10890e5611b", "version": "10.0.0-alpha.1.24431.2" },
  "sdk": { "commit": "9de571282093d76401a3f457a8e9f4427e088778", "version": "10.0.100-alpha.1.24480.5" }
}
```

**Comment by marcpopMSFT** [comment]
> Update: nvm, it is now updated to [dotnet/runtime@0372b50](https://github.com/dotnet/runtime/commit/0372b5080a4aec8d73d6861ad4b993d9d0fe3a11).
We had some issues getting the build actually working. I had been waiting to get the table updated till we actually had a working build: https://github.com/dotnet/sdk/pull/43918

**Pull request merged**